## Program 1

**Write a Python program to print "Hello Python".**

In [1]:
print('Hello python')

Hello python


## Program 2
**Write a Python program to do arithmetical operations addition and division.**

In [2]:
# Addition
num1=eval(input('Enter the value of Num1:'))
num2=eval(input('Enter the value of num2:'))
add=num1+num2
print('THe sum of {} and {} is {}'.format(num1,num2,add))

#Division
num3=eval(input('Enter the value of num1:'))
num4=eval(input('Enter the value of num2:'))
div=num3/num4
print('The division of {} and {} is {}'.format(num3,num4,div))

Enter the value of Num1: 45
Enter the value of num2: 56


THe sum of 45 and 56 is 101


Enter the value of num1: 34
Enter the value of num2: 34


The division of 34 and 34 is 1.0


## Program 3
**Write a Python program to find the area of a triangle.**

In [3]:
base=eval(input('Enter the value of base of a trinagle:'))
height=eval(input('Enter the value of height of a triangle:'))
area=0.5*base*height
print('The area of triangle is {}'.format(area))

Enter the value of base of a trinagle: 45
Enter the value of height of a triangle: 55


The area of triangle is 1237.5


## Program 4
**Write a Python program to swap two variables.**

In [4]:
a=eval(input('Enter the value of a:'))
b=eval(input('Enter the value of b:'))
print('the orginal value of a is {} and b is {}'.format(a,b))
res=a
a=b
b=res
print('the swipe value of a is {} and b is {}'.format(a,b))

Enter the value of a: 45
Enter the value of b: 34


the orginal value of a is 45 and b is 34
the swipe value of a is 34 and b is 45


## Program 5
**Write a Python program to generate a random number.**

In [5]:
import random 
a=random.randint(1,100)
print(a)

73


## Program 6
**Write a Python program to convert kilometers to miles.**

In [6]:
km=eval(input('Enter the distance in km:'))
miles=km*0.621371
print('The {} km is {} in miles'.format(km,miles))

Enter the distance in km: 45


The 45 km is 27.961695 in miles


## Program 7
**Write a Python program to convert Celsius to Fahrenheit.**

In [7]:
celsius=eval(input('Enter the temperature in celsius:'))
fah=(celsius*9/5) + 32
print('The {} degree celsius is in {} degree fahreinheit'.format(celsius,fah))

Enter the temperature in celsius: 45


The 45 degree celsius is in 113.0 degree fahreinheit


## Program 8
**Write a Python program to display calendar.**

In [8]:
import calendar
year=eval(input('Enter the Year:'))
month=eval(input('Enter the month'))
cal=calendar.month(year,month)
print(cal)

Enter the Year: 45
Enter the month 12


    December 45
Mo Tu We Th Fr Sa Su
             1  2  3
 4  5  6  7  8  9 10
11 12 13 14 15 16 17
18 19 20 21 22 23 24
25 26 27 28 29 30 31



## Program 9
**Write a Python program to solve quadratic equation.**

In [9]:
import math
a=eval(input('Enter the value of a:'))
b=eval(input('Enter the value of b:'))
c=eval(input('Enter the value of c:'))

d=b**2-(4*a*c)

if d>0:
    r1=(-b+math.sqrt(d))/(2*a)
    r2=(-b-math.sqrt(d))/(2*a)
    print('The root1 is {}'.format(r1))
    print('The root2 is {}'.format(r2))
elif d==0:
    r3= -b/2*a
    print('The root3 is {}'.format(r3))

else:
    real=-b/2*a
    img=math.sqrt(abs(d))/(2*a)
    print(f"Root 1: {real} + {img}i")
    print(f"Root 2: {real} - {img}i")

Enter the value of a: 33
Enter the value of b: 45
Enter the value of c: 5


The root1 is -0.12203172115420705
The root2 is -1.2416046424821565


## Program 10
**Write a Python program to swap two variables without temp variable.**

In [10]:
a=10
b=50

a,b=b,a

print('After Swapping:')
print('a=',a)
print('b=',b)

After Swapping:
a= 50
b= 10


## Program 11
**Write a Python Program to Check if a Number is Positive, Negative or Zero.**

In [11]:
num=eval(input('Enter the value of num:'))
if num>0:
    print('It is a positive number')
elif num==0:
    print('It is a Zero number.')
else:
    print('it is a negative number.')
    
    

Enter the value of num: 34


It is a positive number


## Program 12
**Write a Python Program to Check if a Number is Odd or Even.**

In [12]:
num=eval(input('Enter the value of num:'))
if num%2==0:
    print('The number {} is an even number'.format(num))
else:
    print('The number {} is an odd number'.format(num) )

Enter the value of num: 56


The number 56 is an even number


## Program 13
**Write a Python Program to Check Leap Year.**

In [13]:
year=eval(input('Enter the year:'))

if (year%400==0) and (year % 100==0) and (year % 100 !=0):
    print('{} is a leap year'.format(year))
else:
    print('{} is not a leap year'.format(year))
    
    

Enter the year: 2019


2019 is not a leap year


## Program 14
**Write a Python Program to Check Prime Number.**

In [16]:
c=0
num=eval(input('Enter the value of num:'))
for i in range(2,num//2):
    c +=1
if c==0:
    print('Prime')
else:
    print('Not prime')


Enter the value of num: 67


Not prime


## Program 15 ¶
**Write a Python Program to Print all Prime Numbers in an Interval of 1-50.**

In [35]:
count=0
lower=1
upper=500

for num in range(lower,upper+1):
    if num>1:
        for i in range(2,num):
            if (num%i) == 0:
               # count=count+1
                break
        else:
            print(num,end=" ")
            count=count+1
print('The total prime number is {}'.format(count))

2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 101 103 107 109 113 127 131 137 139 149 151 157 163 167 173 179 181 191 193 197 199 211 223 227 229 233 239 241 251 257 263 269 271 277 281 283 293 307 311 313 317 331 337 347 349 353 359 367 373 379 383 389 397 401 409 419 421 431 433 439 443 449 457 461 463 467 479 487 491 499 The total prime number is 95


## Program 16
**Write a Python Program to Find the Factorial of a Number.**

In [39]:
num=eval(input('Enter the number:'))
factorial=1
if num<=0:
    print('Factorial does not exist.')
elif num==0:
    print('The factorial is 1.')
else:
    for i in range(1,num+1):
        factorial=factorial*i
    print('The factorial of {} is {}'.format(num,factorial))

Enter the number: 6


The factorial of 6 is 720


## Program 17
**Write a Python Program to Display the multiplication Table.**

In [40]:
num=eval(input('Enter the value of num:'))
for i in range(1,11):
   print("The multiplication of {} * {} is {}".format(num,i,num*i)) 

Enter the value of num: 34


The multiplication of 34 * 1 is 34
The multiplication of 34 * 2 is 68
The multiplication of 34 * 3 is 102
The multiplication of 34 * 4 is 136
The multiplication of 34 * 5 is 170
The multiplication of 34 * 6 is 204
The multiplication of 34 * 7 is 238
The multiplication of 34 * 8 is 272
The multiplication of 34 * 9 is 306
The multiplication of 34 * 10 is 340


## Program 18
**Write a Python Program to Print the Fibonacci sequence.**

In [43]:
def fibonacci(n):
    fib_series = [0, 1]  # Initialize the series with the first two Fibonacci numbers
    
    # Generate Fibonacci sequence up to n terms
    while len(fib_series) < n:
        fib_series.append(fib_series[-1] + fib_series[-2])
    
    return fib_series

# Example usage: Print Fibonacci series up to 10 terms
n_terms = 15
print("Fibonacci series up to", n_terms, "terms:")
print(fibonacci(n_terms))

Fibonacci series up to 15 terms:
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377]


## Program 19
**Write a Python Program to Check Armstrong Number?**

In [1]:
def is_armstrong(number):
    # Calculate the number of digits in the given number
    num_digits = len(str(number))
    
    # Initialize sum
    sum_of_digits = 0
    temp = number
    
    # Calculate the sum of nth power of each digit
    while temp > 0:
        digit = temp % 10
        sum_of_digits += digit ** num_digits
        temp //= 10
    
    # Check if the number is an Armstrong number
    if number == sum_of_digits:
        return True
    else:
        return False

# Example usage
number = int(input("Enter a number to check if it's an Armstrong number: "))
if is_armstrong(number):
    print(number, "is an Armstrong number")
else:
    print(number, "is not an Armstrong number")


Enter a number to check if it's an Armstrong number:  9474


9474 is an Armstrong number


## Program 20
**Write a Python Program to Find Armstrong Number in an Interva**.

In [1]:
lower=eval(input('Enter the lower limit:'))
upper=eval(input('Enter the upper limitL'))

for num in range(lower,upper+1):
    order=len(str(num))
    temp_num= num
    sum=0
    while temp_num>0:
        digit=temp_num%10
        sum+=digit**order
        temp_num//=10

    if num==sum:
        print(num)

Enter the lower limit: 10
Enter the upper limitL 1000


153
370
371
407


## Program 21
**Write a Python Program to Find the Sum of Natural Numbers.**

In [4]:
digit=eval(input('Enter the total digit number:'))
sum=0
for  i in range(1,digit+1):
    sum=sum+i

print("The sum of natural number upto {} is {}".format(digit,sum))

Enter the total digit number: 50


The sum of natural number upto 50 is 1275


## Program 22
**Write a Python Program to Find LCM.**

In [5]:
def lcm(x,y):
    if x>y:
        grater=x
    else:
        greater=y
    while(True):
        if ((greater% x==0)and(greater % y==0)):
            lcm=greater
            break
        greater=greater+1
    return lcm
num1=int(input("Enter the first number:"))
num2=int(input('Enter the second number:'))
print("The L.C.M is",lcm(num1,num2))

Enter the first number: 45
Enter the second number: 54


The L.C.M is 270


## Program 23
**Write a Python Program to Find HCF.**

In [6]:
def hcf(x,y):
    if x>y:
        smaller=y
    else:
        smaller=y
    for i in range(1,smaller+1):
        if((x%i==0)and (y%i==0)):
            hcf=i
    return hcf
num1=int(input("Enter the first number:"))
num2=int(input("Enter the second number:"))

print("The H.C.F is",hcf(num1,num2))

Enter the first number: 54
Enter the second number: 24


The H.C.F is 6


## Program 24
**Write a Python Program to Convert Decimal to Binary, Octal and Hexadecimal.**

In [8]:
num=eval(input('Enter the value of number:'))

print("The decimal value is",num,"is:")

print(bin(num),'in binary.')
print(oct(num),"in octal.")
print(hex(num),"in hexadecimal.")

Enter the value of number: 27


The decimal value is 27 is:
0b11011 in binary.
0o33 in octal.
0x1b in hexadecimal.


## Program 25
**Write a Python Program To Find ASCII value of a character.**

In [11]:
character=str(input("Enter the character:"))
print("The Ascii value of "+character+"is",ord(character) )
                  

Enter the character: P


The Ascii value of Pis 80


## Program 26
**Write a Python Program to Make a Simple Calculator with 4 basic mathematical
operations.**

In [31]:
def add(x,y):
    return x+y

def sub(x,y):
    return x-y

def mul(x,y):
    return x*y

def div(x,y):
    return x/y

print("select operation.")
print("1.Add")
print("2.SUbstract")
print("3.Multiplication")
print("4.Division")
while True:
    choice=input("Enter choice (1/2/3/4):")

    if choice in ('1','2','3','4'):
        try:
            num1=eval(input('Enter the value of x:'))
            num2=eval(input('Enter the value of y:'))
        except ValueError:
            print('Invalid input, Please Enter a number:')
            continue
        if choice =='1':
            print(num1,'+',num2,'=',add(num1,num2))
        elif choice == '2':
           print(num1,'-',num2,'=',sub(num1,num2))
        elif choice =='3':
           print(num1,'*',num2,'=',mul(num1,num2))
        elif choice =='4':
           print(num1,'/',num2,'=',div(num1,num2))
        next_cal=input("Let's do next calculation ? (yes/No)")
        if next_cal=="No":
            break
    else:
        print('Invalid Input')
        
            


select operation.
1.Add
2.SUbstract
3.Multiplication
4.Division


Enter choice (1/2/3/4): 3
Enter the value of x: 23
Enter the value of y: 23


23 * 23 = 529


Let's do next calculation ? (yes/No) yes
Enter choice (1/2/3/4): 5


Invalid Input


Enter choice (1/2/3/4): 2
Enter the value of x: 5000
Enter the value of y: 67


5000 - 67 = 4933


Let's do next calculation ? (yes/No) no
Enter choice (1/2/3/4): 2
Enter the value of x: 34
Enter the value of y: 3


34 - 3 = 31


Let's do next calculation ? (yes/No) No


## Program 27
**write a python program to Display Fibonacci Sequence Using Recurssion**

In [ ]:
def fibonacci(n):
    if n <= 1:
        return n
    else:
        return fibonacci(n-1) + fibonacci(n-2)

def display_fibonacci_sequence(count):
    if count <= 0:
        print("Please enter a positive integer.")
    else:
        print("Fibonacci sequence:")
        for i in range(count):
            print(fibonacci(i), end=" ")


count = int(input("Enter the number of terms in the Fibonacci sequence: "))
display_fibonacci_sequence(count)


Enter the number of terms in the Fibonacci sequence:  90


Fibonacci sequence:
0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 6765 10946 17711 28657 46368 75025 121393 196418 317811 514229 832040 1346269 2178309 3524578 5702887 9227465 14930352 24157817 39088169 63245986 102334155 165580141 267914296 433494437 701408733 1134903170 

## program 28
**Write a python Program to Find Factorial of number Using Recursion**

In [ ]:
def factorial(n):
    if n==1:
        return 1
    else:
        return n*factorial(n-1)
num=int(input('Enter the number:'))

if num<0:
    print('Factorial Does not exist.')

elif num == 0:
    print('Factorial is 1.')
else:
    print('The factorial of {} is {}'.format(num,factorial(n)))

## Program 29
**Write a python program to calculate your Body mass index**

In [4]:
def bodymass(height , weight):
    return round((weight / height**2),2)

h = eval(input('Enter the height in meter:'))
w = eval(input('Enter the weight in kg:'))

print('Welcome to the bmi calculator:')
bmi=bodymass(h,w)
print('your BMI is:',bmi)

if bmi<=18.5:
    print('you are underweight.')
elif 18.5< bmi <=24.9:
    print('you are normal weight')
elif 25<bmi<29.29:
    print('You are overweight.')
else:
    print('You are obese')

Enter the height in meter: 1.7
Enter the weight in kg: 56


Welcome to the bmi calculator:
your BMI is: 19.38
you are normal weight


## program 30

**Write a python program to caculate the natural logarithm of any number**

In [8]:
import math
num=eval(input("Enter a number:"))

if num<=0:
    
    print('please enter a positive number.')
else:
    res=math.log(num)
    print("the natural algorithm of {} is {}".format(num,res))


Enter a number: 10


the natural algorithm of 10 is 2.302585092994046


## Program 31
**Write a python program for cube sum of first n natural number**

In [10]:
def cube(n):
    if n<=0:
        return 0
    else:
        total=sum([i**3 for i in range(1,n+1)])
        return total

n= eval(input("Enter the value of n:"))
if n<=0:
    print('please enter a positive number.')
else:
    res=cube(n)
    print('The cube sum of first {} is {}'.format(num,res))

Enter the value of n: 10


The cube sum of first 10 is 3025


## Program 32

**Write a python program to find sum of array**

In [11]:
arr=[1,2,3]
ans=sum(arr)
print('Sum of the array is :',ans)

Sum of the array is : 6


In [13]:
def sum_of_array(arr):
    total=0

    for element in arr:
        total +=element

    return total

array=[1,2,3,4,5,6]
res=sum_of_array(array)
print("Sum of the array:",res)

Sum of the array: 21


## Program 33
**Write a python program to find largest element in an array**

In [1]:
def largest(arr):
    if not arr:
        return 'Array is empty'

    largest_number=arr[0]

    for element in arr:
        if element>largest_number:
            largest_number=element

    return largest_number

array=[100,230,4444,55550,6754,2044359594]
result=largest(array)
print('The largest elemnt in the array is {}'.format(result))

The largest elemnt in the array is 2044359594


## Program 34
**Write a Python Program for array rotation.**

In [4]:
def rotate_array(arr, k):
    """
    Rotate the array 'arr' to the right by 'k' steps.
    """
    n = len(arr)
    k = k % n  # Ensure k is within the range of array length

    # Rotate the array using slicing
    arr[:] = arr[-k:] + arr[:-k]

# Example usage:
if __name__ == "__main__":
    arr = [1, 2, 3, 4, 5]
    k = 2
    print("Original array:", arr)
    rotate_array(arr, k)
    print("Array rotated by", k, "steps:", arr)


Original array: [1, 2, 3, 4, 5]
Array rotated by 2 steps: [4, 5, 1, 2, 3]


In [24]:
def rotate(arr,k):
    n=len(arr)
    k=k%n

    arr[:]=arr[-k:]+arr[:-k]

if __name__=="__main__":
    arr=[4575,7348734658,32778348,7273,28724,28,23,211,345,343,4312,23432]
    k=3
    print('The original array is:',arr)
    rotate(arr,k)
    print('Tehe rotated array by',k,'steps:',arr)



The original array is: [4575, 7348734658, 32778348, 7273, 28724, 28, 23, 211, 345, 343, 4312, 23432]
Tehe rotated array by 3 steps: [343, 4312, 23432, 4575, 7348734658, 32778348, 7273, 28724, 28, 23, 211, 345]


## Program 35
**Write a Python Program to Split the array and add the first part to the end?**

In [27]:
def split(arr,k):
    if k<=0 or k>=len(arr):
        return arr

    first=arr[:k]
    second=arr[k:]
    res=second+first
    return result

arr=[1,2,3,4,5,6,7,8,9]
k=4
res=split(arr,k)
print('Original Array is',arr)
print("The split and add array is",res)

Original Array is [1, 2, 3, 4, 5, 6, 7, 8, 9]
The split and add array is 2044359594


## Program 36
**Write a Python program to check if given array is monotonic**

In [29]:
def monotonic(arr):
    increasing=decreasing=True

    for i in range(1,len(arr)):
        if arr[i]>arr[i-1]:
            decreasing = False
        elif arr[i]<arr[i-1]:
            increasing = False
    return increasing or decreasing


arr1=[1,2,2,3,3,44,5]
arr2=[4,5,9,1,2,3]
arr3=[3,2,1]

print('arr1 is monotic:',monotonic(arr1))
print('arr2 is monotic:',monotonic(arr2))
print('arr3 is monotic:',monotonic(arr3))

arr1 is monotic: False
arr2 is monotic: False
arr3 is monotic: True


## program 37
**Write a python program to add two metrices**

In [32]:
def add_metrices(mat1,mat2):
    if len(mat1)!=len(mat2) or len(mat1[0])!=len(mat2[0]):
        return 'metrices must have same shape and size for addition'

    result=[]
    for i in range(len(mat1)):
        row=[]
        for j in range(len(mat1[0])):
            row.append(mat1[i][j]+mat2[i][j])
        result.append(row)

    return result

matrix1=[
    [1,2,3],
    [4,5,6],
    [7,8,9]
]

matrix2=[
    [9,8,7],
    [6,5,4],
    [3,2,1]
]

result_matrix=add_metrices(matrix1,matrix2)

if isinstance(result_matrix,str):
    print(result_metrices)

else:
    print("Sum of metrices:")
    for row in result_matrix:
        print(row)

Sum of metrices:
[10, 10, 10]
[10, 10, 10]
[10, 10, 10]


## program 38
**Write a python program to multiply two metrices**

In [39]:
def multiply_matrices(mat1,mat2):
    rows1=len(mat1)
    cols1=len(mat1[0])
    rows2=len(mat2)
    cols2=len(mat2[0])

    if cols1 != rows2:
        return "Matrix multiplication is not possible."
    result = [[0 for _ in range(cols2)] for _ in range(rows1)]

    for i in range(rows1):
        for j in range(cols2):
            for k in range (cols1):
                result[i][j] +=mat1[i][k] * mat2[k][j]

    return result
matrix1 = [[1,2,3],
           [4,5,6]]

matrix2=[[7,8],
         [9,10],
         [11,12]]

result_matrix = multiply_matrices(matrix1,matrix2)

if isinstance(result_matrix,str):
    print(result_matrix)
else:
    print("Result of matrix multiplication:")
    for row in result_matrix:
        print(row)
    

Result of matrix multiplication:
[58, 64]
[139, 154]


## program 39
**Write a python program to Transpose a Matrix**

In [44]:
def transpose_matrices(matrix):
    rows,cols = len(matrix),len(matrix[0])

    result = [[0 for _ in range(rows)] for _ in range(cols)]

    for i in range(rows):
        for j in range(cols):
            result[j][i] = matrix[i][j]
    return result

matrix=[
    [1,2,3],
    [4,5,6]
    
]

transpose_matrix=transpose_matrices(matrix)

for row in transpose_matrix:
    print(row)

[1, 4]
[2, 5]
[3, 6]


## program 40
**Write a python Program to sort words in Alphabetic Order**

In [49]:
my_str = input("Enter a string: ")

words=[word.capitalize() for word in my_str.split()]

words.sort()

print("The sorted words are:")
for words in words:
    print(words)

Enter a string:  ramesh suresh mahesh rajua nisha 


The sorted words are:
Mahesh
Nisha
Rajua
Ramesh
Suresh


## Program 41

**Write a python program to Remove Punctuation From a String**

In [57]:
punctuations='!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'
my_str=input("Enter a string:")

no_punct = ""
for char in my_str:
    if char not in punctuations:
        no_punct=no_punct+char

print(no_punct)


Enter a string: @#$Deepak heeloo Good morning How !###$ How Are you &#@Y^


Deepak heeloo Good morning How  How Are you Y


## Program 43
**Write a python program to check if the given number is Disarium Number .**

In [64]:
def is_disarium(number):
    num_str = str(number)

    digit_sum=sum(int(i) ** (index+1) for index, i in enumerate(num_str))
    return digit_sum == number

try:
    num = int(input("Enter a number:"))
    if is_disarium(num):
        print(f"{num} is a Disarium number.")
    else:
        print(f"{num} is not a Disarium number.")
except ValueError:
    print("Invalid input. please enter a valid number")

Enter a number: 89


89 is a Disarium number.


## Program 44
**Write a python program to check all Diarium number between 1 to 100**

In [65]:
def is_disarium(num):
    num_str=str(num)
    digit_sum=sum(int(i) ** (index+1) for index,i in enumerate(num_str))
    return num == digit_sum

disarium_numbers = [num for num in range(1,101) if is_disarium(num)]

print("Disarium numbers between 1 and 100:")
for num in disarium_numbers:
    print(num,end='||')

Disarium numbers between 1 and 100:
1||2||3||4||5||6||7||8||9||89||

## Program 45
**Write a python Program to check if the given number is Happy Number**

In [71]:
def is_happy_number(num):
    seen=set()

    while num!=1 and num not in seen:
        seen.add(num)
        num=sum(int(i) ** 2 for i in str(num))

    return num==1

num=eval(input('Enter the value of number:'))
if is_happy_number(num):
    print("{} is a happy number".format(num))
else:
    print("{} is not a Happy number.".format(num))

Enter the value of number: 32


32 is a happy number


## Program 46
**Write a python program to print all happy numbers between 1 to 100**

In [72]:
def is_happy_number(num):
    seen=set()

    while num !=1 and num not in seen:
        seen.add(num)
        num=sum(int(i) ** 2 for i in str(num))

    return num==1

happy_numbers=[]
for num in range(1,101):
    if is_happy_number(num):
        happy_numbers.append(num)
print('Happy number between 1 to 100:')
print(happy_numbers)

Happy number between 1 to 100:
[1, 7, 10, 13, 19, 23, 28, 31, 32, 44, 49, 68, 70, 79, 82, 86, 91, 94, 97, 100]


## Program 47
**Write a Python program to determine whether the given number is a Harshad
Number.**

In [3]:
def is_harshad_number(num):
    digit_sum = sum(int(i) for i in str(num))

    return num%digit_sum ==0
num=eval(input("Enter the number:"))

if is_harshad_number(num):
    print("The {} is harshad number".format(num))
else:
    print('The {} is not a harshad number'.format(num))

Enter the number: 234


The 234 is harshad number


## Program 48
**Write a Python program to print all pronic numbers between 1 and 100.**

In [6]:
def pronic_number(num):
    for n in range(1,int(num**0.5)+1):
        if n*(n+1) == num:
            return True
    return False

print("Protonic numbers between 1 to 100 are:")
for i in range(1,101):
     if pronic_number(i):
         print(i,end=' | ')

Protonic numbers between 1 to 100 are:
2 | 6 | 12 | 20 | 30 | 42 | 56 | 72 | 90 | 

## Program 49
**Write a python program to find sum of element in a list**

In [7]:
l1=[100,200,340,560]
sum=0
for i in l1:
    sum=sum+i

print("The sum of all the element in a list is {}".format(sum))

The sum of all the element in a list is 1200


## Program 50
**Write a Python program to Multiply all numbers in the list.**

In [9]:
l1=[10,20,30,5]
product=1

for i in l1:
    product*=i

print("The product of all number is a list is {}".format(product))

The product of all number is a list is 30000


## Program 51
**Write a Python program to find smallest number in a list.**

In [12]:
l1=[10,20,30,5,45,6,2,1,3,0,-3]

minnimum = l1[0]

for i in l1:
    if i < minnimum:
        minnimum=i

print('The minnimum number is {}'.format(minnimum))
    

The minnimum number is -3


## Program 52
**Write a Python program to find largest number in a list.**

In [13]:
l1=[10,20,30,5,45,6,2,1,3,0,-3]

maximum=l1[0]

for i in l1:
    if i>maximum:
        maximum=i

print('The maximim number is {} '.format(maximum))

The maximim number is 45 


## Program 53
**Write a Python program to find second largest number in a list.**

In [16]:
l1=[10,20,30,5,45,6,2,1,3,0,-3]
l1.sort(reverse=True)

if len(l1)>=2:
    second_largest = l1[1]
    print('The second largest number in the list is {}:'.format(second_largest))

else:
    ('The list doesnot contain the second largest number')


The second largest number in the list is 30:


## Program 54
**Write a Python program to find N largest elements from a list.**

In [17]:
def find_n_largest(list1,n):
    sorted_list1=sorted(list1,reverse=True)

    largest_elements = sorted_list1[:n]

    return largest_elements
numbers = [30, 10, 45, 5, 20, 50, 15, 3, 345, 54, 67, 87, 98, 100, 34,76847567,]
N=int(input("N="))

result = find_n_largest(numbers,N)


print("The largest element in the is {}".format(N),result)

N= 4


The largest element in the is 4 [76847567, 345, 100, 98]


## Program 55
**Write a Python program to print even numbers in a list.**

In [18]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

even=[num for num in numbers if num%2==0]

print("Even number is list are",even)

Even number is list are [2, 4, 6, 8, 10]


## Program 56
**Write a Python program to print odd numbers in a List.**

In [19]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

odd=[num for num in numbers if num%2!=0]

print("Even number is list are",odd)

Even number is list are [1, 3, 5, 7, 9]


## Program 57
**Write a Python program to Remove empty List from List.**

In [20]:
list_of_lists = [[1, 2, 3], [], [4, 5], [], [6, 7, 8], []]
filter=[i for i in list_of_lists if i]

print('List after removing empty lists:',filter)

List after removing empty lists: [[1, 2, 3], [4, 5], [6, 7, 8]]


## Program 58
**Write a Python program to Cloning or Copying a list.**

In [23]:
or_list=[1,2,3,4,5]
clone_list=or_list[:]
print(clone_list)

[1, 2, 3, 4, 5]


In [24]:
or_list=[1,2,3,4,5]
clone_list=list(or_list)
print(clone_list)

[1, 2, 3, 4, 5]


In [25]:
or_list=[1,2,3,4,5]
clone_list=[item for item in or_list]
print(clone_list)

[1, 2, 3, 4, 5]


## Program 59
**Write a Python program to Count occurrences of an element in a list.**

In [29]:
def count(l,element):
    count=l.count(element)
    return count

my_list = [1, 2, 3, 4, 2, 5, 2, 3, 4, 6, 5,2,2,2,3,4,5,2,3,2,32,32,3,2]

element_count=2
occurences = count(my_list , element_count)

print("the element {} appears {} times in a list".format(element_count,occurences))

the element 2 appears 9 times in a list


## Program 60
**Write a Python program to find words which are greater than given length k.**

In [31]:
def find_word(words,k):
    result=[]
    for i in words:
        if len(i)>k:
            result.append(i)
    return result

word_list=["apple","banana", "cherry", "date", "elderberry", "dragonfruit","Deepak","Bhutaki"]

k=5
long_words=find_word(word_list,k)
print(f"Words longer than {k} characters: {long_words}")

Words longer than 5 characters: ['banana', 'cherry', 'elderberry', 'dragonfruit', 'Deepak', 'Bhutaki']


## Program 61
**Write a Python program for removing i character from a string. th**

In [34]:
def remove_char(input_str,i):
    if i<0 or i> len(input_str):
        print("Invalid index {}.the string remains unchanged".format(i))
        return result_str

    result_str=input_str[:i]+input_str[i+1:]

    return result_str
input_str="Hello ,wWorld!"
i=7
new_str=remove_char(input_str,i)
print(f"Original String: {input_str}")
print(f"String after removing {i}th character : {new_str}")

Original String: Hello ,wWorld!
String after removing 7th character : Hello ,World!


## Program 62
**Write a python program to split and join a sstring**

In [36]:
input_str="python program to split and join a string"
word_list=input_str.split()
seperator= " "
output_str=seperator.join(word_list)


print("Original string:",input_str)
print("List of solit words:",word_list)
print("Joined string:",output_str)

Original string: python program to split and join a string
List of solit words: ['python', 'program', 'to', 'split', 'and', 'join', 'a', 'string']
Joined string: python program to split and join a string


## Program 63
**Write a python program to check if a given string is binary or not**

In [39]:
def is_binary(input_str):
    for i in input_str:
        if i not in '01':
            return False
    return True

input_str="011000111000"

if is_binary(input_str):
    print("{} is a binary string.".format(input_str))
else:
    print("{} is not a binary string".format(input_str))

011000111000 is a binary string.


## Program 64
**Write a Python program to find uncommon words from two Strings.**

In [40]:
def uncommon_words(str1,str2):
    words1=set(str1.split())
    words2=set(str2.split())

    uncommon_words_set=words1.symmetric_difference(words2)

    uncommon_words_list=list(uncommon_words_set)
    return uncommon_words_list


string1 = "This is the first string"
string2 = "This is the second string"
uncommon =  uncommon_words(string1,string2)

print("Uncommon words:",uncommon)

Uncommon words: ['second', 'first']


## Program 65
**Write a Python program to find all duplicate characters in string.**

In [41]:
def find_duplicates(input_str):
    char_count={}
    duplicates=[]

    for i in input_str:
        if i in char_count:
            char_count[i]+=1
        else:
            char_count[i]=1
    for i,count in char_count.items():
        if count>1:
            duplicates.append(i)

    return duplicates

input_string="Deepak kumar"

duplicate_chars = find_duplicates(input_string)

print("Duplicates characters:",duplicate_chars)

Duplicates characters: ['e', 'a', 'k']


## Program 66
**Write a Python Program to check if a string contains any special character.**

In [7]:
import re

def chech_special_char(in_str):
    pattern = r'[!@#$%^&*()_+{}\[\]:;<>,.?~\\\/\'"\-=]'

    if re.search(pattern,in_str):
         return True
    else:
        return False
input_string= str(input("Enter a string:"))
contains_special=chech_special_char(input_string)

if contains_special:
    print("The string contains special characters.")
else:
    
    print("The string does not contain special character")



Enter a string: "hello deepak , how are you ?"


The string contains special characters.


## Program 67
**Write a Python program to Extract Unique dictionary values.**

In [2]:
my_dict={
    'a':10,
    'b':20,
    'c':30,
    'd':40,
    'e':20,
    'f':40,
}
uni_value=set()

for i in my_dict.values():
    uni_value.add(i)

unique_values_list = list(uni_value)

print('unique values in the dictionary is :', unique_values_list)

unique values in the dictionary is : [40, 10, 20, 30]


## Program 68
**Write a Python program to find the sum of all items in a dictionary.**

In [3]:
my_dict={
    'a':10,
    'b':20,
    'c':30,
    'd':40,
    'e':20,
    'f':40,
}

total_sum=0
for i in my_dict.values():
    total_sum +=i

print("Sum of all items in the dictionary is :",total_sum)

Sum of all items in the dictionary is : 160


## Program 69
**Write a Python program to Merging two Dictionaries.**

In [4]:
dict1 = {'a': 1, 'b': 2}
dict2 = {'c': 3, 'd': 4}

dict1.update(dict2)

print("merged Dictioanry is",dict1)

merged Dictioanry is {'a': 1, 'b': 2, 'c': 3, 'd': 4}


In [5]:
dict2.update(dict1)

print("merged Dictioanry is",dict2)

merged Dictioanry is {'c': 3, 'd': 4, 'a': 1, 'b': 2}


## Program 70
**Write a Python program to convert key-values list to flat dictionary.**

In [7]:
key_values_list = [('a', 1), ('b', 2), ('c', 3), ('d', 4)]

flat_dict = {}
for key, value in key_values_list:
    flat_dict[key] = value

print("Flat Dictionary :",flat_dict)

Flat Dictionary : {'a': 1, 'b': 2, 'c': 3, 'd': 4}


## Program 71
**Write a Python program to insertion at the beginning in OrderedDict.**

In [10]:
from collections import OrderedDict

ordered_dict = OrderedDict([('b', 2), ('c', 3), ('d', 4)])

new_item = ('a' , 1)
new_ordered_dict = OrderedDict([new_item])

new_ordered_dict.update(ordered_dict)

print("Updated OrderedDict:", new_ordered_dict)

Updated OrderedDict: OrderedDict([('a', 1), ('b', 2), ('c', 3), ('d', 4)])


## Program 72
**Write a Python program to check order of character in string using OrderedDict().**

In [13]:
from collections import OrderedDict

def check_order(string,reference):
    string_dict = OrderedDict.fromkeys(string)
    reference_dict=OrderedDict.fromkeys(reference)

    return string_dict==reference_dict

input_string="hello world"
refernce_string="helo wrd"

if check_order(input_string,refernce_string):
    print("The order of character in the input string matches the refernce string",input_string)

else:
    print("The order of character in the input string does not match the reference string",reference_string)

The order of character in the input string matches the refernce string hello world


## Program 73
**Write a Python program to sort Python Dictionaries by Key or Value.**

In [15]:
sample_dict= {'appple':3,'banana':1,'cherry':2,'date':4}

sorted_dict_by_keys = dict(sorted(sample_dict.items()))

print("Sorted by keys:")
for key,value in sorted_dict_by_keys.items():
    print(f"{key}:{value}")

Sorted by keys:
appple:3
banana:1
cherry:2
date:4


## Program 74
**Write a program that calculates and prints the value according to the given formula:**
Q = square root of 2CD/H

Following are the fixed values of C and H:

C is 50. H is 30.

D is the variable whose values should be input to your program in a commaseparated sequence.

Example

Let us assume the following comma separated input sequence is given to the
program:

100,150,180

The output of the program should be:

18,22,24


In [16]:
import math
C = 50
H = 30

def calculate_Q(D):
    return int(math.sqrt((2 * C * D)/H))

input_sequence = input("Enter comma-seperated values of D:")
D_values = input_sequence.split(',')

result = [calculate_Q(int(D)) for D in D_values]
print(','.join(map(str,result)))

Enter comma-seperated values of D: 21348,8653,6235


266,169,144


## Program 75
**Write a program which takes 2 digits, X,Y as input and generates a 2-dimensional
array.**
The element value in the i-th row and j-th column of the array should be i*j.

Note: i=0,1.., X-1; j=0,1,¡Y-1.

Example

Suppose the following inputs are given to the program:

3,5

Then, the output of the program should be:

[[0, 0, 0, 0, 0], [0, 1, 2, 3, 4], [0, 2, 4, 6, 8]]

In [17]:
X,Y = map(int,input("Enter two digits (X,Y): ").split(','))

array =[[0 for j in range(Y)] for i in range(X)]

for i in range(X):
    for j in range(Y):
        array[i][j] = i*j

for row in array:
    print(row)

Enter two digits (X,Y):  3,5


[0, 0, 0, 0, 0]
[0, 1, 2, 3, 4]
[0, 2, 4, 6, 8]


## Program 76
**Write a program that accepts a comma separated sequence of words as input and
prints the words in a comma-separated sequence after sorting them alphabetically.**
Suppose the following input is supplied to the program:

without,hello,bag,world

Then, the output should be:

bag,hello,without,world


In [19]:
input_sequence = input("Enter a comma-seperated sequence of words:")

words= input_sequence.split(',')

sorted_words = sorted(words)

sorted_sequence = ','.join(sorted_words)

print("Sorted words:",sorted_sequence)

Enter a comma-seperated sequence of words: without,hello,bag,world


Sorted words: bag,hello,without,world


## Program 77
**Write a program that accepts a sequence of whitespace separated words as input
and prints the words after removing all duplicate words and sorting them
alphanumerically.**

Suppose the following input is supplied to the program:

hello world and practice makes perfect and hello world again

Then, the output should be:

again and hello makes perfect practice world


In [21]:
input_sequence = input("Enter a sequence of whitespace-seperated words:")

words = set(input_sequence.split())

sorted_words = sorted(words)

result= ''.join(sorted_words)

print("Result:",result)

Enter a sequence of whitespace-seperated words: hello world and practice makes perfect and hello world again


Result: againandhellomakesperfectpracticeworld


## Program 79
**Write a program that accepts a sentence and calculate the number of letters and
digits. Suppose the following input is supplied to the program:**
    
hello world! 123

Then, the output should be:

LETTERS 10

DIGITS 3


In [22]:
sentence = input('Enter a sentence:')

letter_count=0
digit_count=0

for char in sentence:
    if char.isalpha():
        letter_count +=1
    elif char.isdigit():
        digit_count +=1

print('LETTERS',letter_count)
print('DIGITS',digit_count)

Enter a sentence: hello worldf! 123


LETTERS 11
DIGITS 3


## Program 80
**A website requires the users to input username and password to register. Write a
program to check the validity of password input by users. Following are the criteria
for checking the password:**

**1. At least 1 letter between [a-z]**

**2. At least 1 number between [0-9]**

**1. At least 1 letter between [A-Z]**

**3. At least 1 character from [$#@]**

**4. Minimum length of transaction password: 6**

**5. Maximum length of transaction password: 12**


**Your program should accept a sequence of comma separated passwords and will
check them according to the above criteria. Passwords that match the criteria are to
be printed, each separated by a comma.**
### Example
**If the following passwords are given as input to the program:**

**ABd1234@1,a F1#,2w3E*,2We3345**

Then, the output of the program should be:

ABd1234@1

In [24]:
import re

def is_valid_password(password):
    if 6<= len(password)<=12:
        if re.match(r"^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[$#@])",password):
            return True
        return False

passwords = input("Enter passwords seperated by commas:").split(',')

valid_passwords= []
for psw in passwords:
    if is_valid_password(psw):
        valid_passwords.append(psw)

print(','.join(valid_passwords))
    

Enter passwords seperated by commas: Deepak@123


Deepak@123


## Program 81
**Define a class with a generator which can iterate the numbers, which are divisible by
7, between a given range 0 and n.**


In [30]:
class DivisibleBySeven():
    def __int__(self,n):
        self.n=n

    def generate_divisible_by_seven(self):
        for num in range(self.n +1):
            if num%7 ==0:
                yield num

In [31]:
n = int(input("Enter your desired range: "))

divisible_by_seven_generator = DivisibleBySeven(n).generate_divisible_by_seven

for num in divisible_by_seven_generator:
    print(num)

Enter your desired range:  45


TypeError: DivisibleBySeven() takes no arguments

## Program 82
**Write a program to compute the frequency of the words from the input. The output
should output after sorting the key alphanumerically. Suppose the following input is
supplied to the program:**
New to Python or choosing between Python 2 and Python 3? Read Python 2 or

Python 3.

Then, the output should be:

2:2

3.:1

3?:1

New:1

Python:5

Read:1

and:1

between:1

choosing:1

or:2

to:1


In [36]:
input_sentence = input('Enter a sentences:')

words= input_sentence.split()

word_freq = {}

for word in words:
    word = word.strip('.,?')
    word = word.lower()
    if word in word_freq:
        word_freq[word] +=1
    else:
        word_freq[word] =1

sorted_words = sorted(word_freq.items())

for word , frequency in sorted_words:
    print(f"{word}:{frequency}")

Enter a sentences: New to Python or choosing between Python 2 and Python 3?


2:1
3:1
and:1
between:1
choosing:1
new:1
or:1
python:3
to:1


## Program 83
**Define a class Person and its two child classes: Male and Female. All classes have a
method "getGender" which can print "Male" for Male class and "Female" for Female
class**


In [44]:
class Person:
    def getGender(self):
        return "Unknown"

class Male(Person):
    def getGender(self):
        return "Male"

class Female(Person):
    def getGender(self):
        return "Female"
        

In [45]:
person = Person()
male = Male()
female = Female()

print(person.getGender())
print(male.getGender())
print(female.getGender())

Unknown
Male
Female


## Program 84
**Please write a program to generate all sentences where subject is in ["I", "You"] and
verb is in ["Play", "Love"] and the object is in ["Hockey","Football"].**

In [46]:
subjects = ['I' , "You"]
verbs = ["Play","Love"]
objects = ["Hockey","Football"]

sentences =[]
for sub in subjects:
    for vrb in verbs:
        for obj in objects:
            sentence = f"{sub} {vrb} {obj}."
            sentences.append(sentence)

for sentence in sentences:
    print(sentence)

I Play Hockey.
I Play Football.
I Love Hockey.
I Love Football.
You Play Hockey.
You Play Football.
You Love Hockey.
You Love Football.


## Program 85
**Please write a program to compress and decompress the string "hello world!hello
world!hello world!hello world!"**

In [49]:
import zlib

string =  "hello world!hello world!hello world!hello world!"

compressed_string = zlib.compress(string.encode())

decompressed_string = zlib.decompress(compressed_string).decode()

print("Original String:", string)
print("Compressed String:", compressed_string)
print("Decompressed String:", decompressed_string)


Original String: hello world!hello world!hello world!hello world!
Compressed String: b'x\x9c\xcbH\xcd\xc9\xc9W(\xcf/\xcaIQ\xcc \x82\r\x00\xbd[\x11\xf5'
Decompressed String: hello world!hello world!hello world!hello world!


## Program 86
**Please write a binary search function which searches an item in a sorted list. The
function should return the index of element to be searched in the list.**


In [53]:
def binary_search (arr,target):
    left,right = 0, len(arr) - 1

    while left <=right:
        mid = (left + right) //2

        if  arr[mid] == target :
            return mid
        elif arr[mid] < target:
            left = mid +1
        else:
            right = mid -1

    return -1

sorted_list = [1, 2, 3, 4, 5, 6, 7, 8, 9]
target_element = 4

result = binary_search (sorted_list, target_element)

if result != -1:
 print(f"Element {target_element} found at index {result}")
else:
 print(f"Element {target_element} not found in the list")


Element 4 found at index 3


## Program 87
**Please write a program using generator to print the numbers which can be divisible
by 5 and 7 between 0 and n in comma separated form while n is input by console.**

Example:
       
If the following n is given as input to the program:

100

Then, the output of the program should be:

0,35,70


In [54]:
def divisible_by_5_and_7(n):
    for num in range(n+1):
        if num%5==0 and num%7==0:
            yield num

try:
    n=int(input("Enter a value for n:"))
    result = divisible_by_5_and_7(n)
    print(','.join(map(str, result)))

except valueError:
    print("Invalid input.Please enter a valid integer for n.")

Enter a value for n: 100


0,35,70


## Program 88
**Please write a program using generator to print the even numbers between 0 and n in
comma separated form while n is input by console.**

Example:
       
If the following n is given as input to the program:

10

Then, the output of the program should be:

0,2,4,6,8,10

In [55]:
def even_numbers(n):
    for num in range (n+1):
        if  num%2==0:
            yield num

try:
    n=int(input("Enter a value for n:"))
    result = even_numbers(n)
    print(','.join(map(str,result)))

except valueError:
    print("Invalid Input. please enter a valid integer for n")

Enter a value for n: 90


0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90


## Program 89
**The Fibonacci Sequence is computed based on the following formula:**
f(n)=0 if n=0

f(n)=1 if n=1

f(n)=f(n-1)+f(n-2) if n>1

**Please write a program using list comprehension to print the Fibonacci Sequence in
comma separated form with a given n input by console.**

Example:

If the following n is given as input to the program:

8

Then, the output of the program should be:

0,1,1,2,3,5,8,13

In [58]:
def fibonacci(n):
 sequence = [0, 1] # Initializing the sequence with the first two F
 [sequence.append(sequence[-1] + sequence[-2]) for _ in range(2, n)]
 return sequence

try:
    n= int(input("Enter a value of n:"))
    result = fibonacci(n)
    print(','.join(map(str,result)))

except:
    print("inavlid input.print enter a valid integer for n")

Enter a value of n: 45


0,1,1,2,3,5,8,13,21,34,55,89,144,233,377,610,987,1597,2584,4181,6765,10946,17711,28657,46368,75025,121393,196418,317811,514229,832040,1346269,2178309,3524578,5702887,9227465,14930352,24157817,39088169,63245986,102334155,165580141,267914296,433494437,701408733


## Program 90
**Assuming that we have some email addresses in the
"username@companyname.com (mailto:username@companyname.com)" format,
please write program to print the user name of a given email address. Both user
names and company names are composed of letters only.**
    
Example:

If the following email address is given as input to the program:

john@google.com (mailto:john@google.com)

Then, the output of the program should be:

john


In [61]:
def extract_username(email):
    parts = email.split('@')
    if len(parts) == 2:
        return parts[0]

    else:
        return "invalid email format"

try:
    email = input('Enter aa email adress:')
    username = extract_username(email)
    print(username)

except valueError:
    print("Invalid input.please enter a valid email adress.")

Enter aa email adress: chikudeepak0@gmail.com


chikudeepak0


## Program 91
**Define a class named Shape and its subclass Square. The Square class has an init
function which takes a length as argument. Both classes have an area function which
can print the area of the shape where Shape's area is 0 by default.**

In [4]:
class shape:
    def __init__(self):
        pass
    def area(self):
        return 0

class square(shape):
    def __init__(self,length):
        super().__init__()
        self.length = length

    def area(self):
        return self.length**2

shape = shape()
square = square(float(input("Enter the shape of the square:")))

print("Shape's area by default:",shape.area())
print("Area of the square:",square.area())

Enter the shape of the square: 5


Shape's area by default: 0
Area of the square: 25.0


## Program 92
**Write a function that stutters a word as if someone is struggling to read it. The first
two letters are repeated twice with an ellipsis ... and space after each, and then the
word is pronounced with a question mark ?.**

Examples

stutter("incredible") ➞ "in... in... incredible?"

stutter("enthusiastic") ➞ "en... en... enthusiastic?"

stutter("outstanding") ➞ "ou... ou... outstanding?"

Hint :- Assume all input is in lower case and at least two characters long.

In [7]:
def stutter(word):
    if len(word)<2:
        return "word must be at least two character long."

    stutter_word = f"{word[:2]}...{word[:2]}...{word}?"
    return stutter_word
print(stutter('Deepak'))
print(stutter('Nisha'))
print(stutter('raghav'))

De...De...Deepak?
Ni...Ni...Nisha?
ra...ra...raghav?


## Program 93
**Create a function that takes an angle in radians and returns the corresponding angle
in degrees rounded to one decimal place.**

Examples

radians_to_degrees(1) ➞ 57.3

radians_to_degrees(20) ➞ 1145.9

radians_to_degrees(50) ➞ 2864.8


In [8]:
import math

def radians_to_degrees(radians):
    degrees = radians*(180/math.pi)
    return round(degrees,1)


print(radians_to_degrees(20))
print(radians_to_degrees(45))
print(radians_to_degrees(100))
print(radians_to_degrees(120))

1145.9
2578.3
5729.6
6875.5


## Program 94
**In this challenge, establish if a given integer num is a Curzon number. If 1 plus 2
elevated to num is exactly divisible by 1 plus 2 multiplied by num, then num is a
Curzon number.**
    
**Given a non-negative integer num, implement a function that returns True if num is a
Curzon number, or False otherwise.**
    
Examples

is_curzon(5) ➞ True
    
2 ** 5 + 1 = 33

 2 * 5 + 1 = 11
 
 33 is a multiple of 11
    
is_curzon(10) ➞ False
    
2 ** 10 + 1 = 1025

 2 * 10 + 1 = 21
 
1025 is not a multiple of 21
    
is_curzon(14) ➞ True
    
2 ** 14 + 1 = 16385

2 * 14 + 1 = 29

16385 is a multiple of 29

**Curzon Number:**
    
**It is defined based on a specific mathematical relationship involving powers of 2. An integer
'n' is considered a Curzon number if it satisfies the following condition:**
    
If (2^n + 1) is divisible by (2n + 1), then 'n' is a Curzon number.

For example:

If n = 5: 2^5 + 1 = 33, and 2*5 + 1 = 11. Since 33 is divisible by 11 (33 % 11 = 0), 5 is a
Curzon number.
    
If n = 10: 2^10 + 1 = 1025, and 2*10 + 1 = 21. 1025 is not divisible by 21, so 10 is not a
Curzon number.
    
Curzon numbers are a specific subset of integers with this unique mathematical property.


In [9]:
def is_curzon(num):
    numerator = 2 ** num +1
    denominator = 2 * num+1
    return numerator%denominator == 0

print(is_curzon(34))
print(is_curzon(45))
print(is_curzon(10))
print(is_curzon(14))
print(is_curzon(5))

False
False
False
True
True


## Program 95
**Given the side length x find the area of a hexagon.**

Examples

area_of_hexagon(1) ➞ 2.6

area_of_hexagon(2) ➞ 10.4

area_of_hexagon(3) ➞ 23.4


In [10]:
import math

def area_of_hexagon(x):
    area = (3* math.sqrt(3)*x**2)/2
    return round(area ,1)

print(area_of_hexagon(3))
print(area_of_hexagon(4))
print(area_of_hexagon(10))

23.4
41.6
259.8


##Program 96
**Create a function that returns a base-2 (binary) representation of a base-10 (decimal)
string number. To convert is simple: ((2) means base-2 and (10) means base-10)
010101001(2) = 1 + 8 + 32 + 128.**

**Going from right to left, the value of the most right bit is 1, now from that every bit to
the left will be x2 the value, value of an 8 bit binary numbers are (256, 128, 64, 32, 16,
8, 4, 2, 1).**

Examples

binary(1) ➞ "1"

1*1 = 1

binary(5) ➞ "101"

1*1 + 1*4 = 5

binary(10) ➞ 1010

1*2 + 1*8 = 10


In [11]:
def binary(decimal):
    binary_str = " "
    while decimal > 0:
        remainder = decimal % 2
        binary_str = str(remainder) + binary_str
        decimal = decimal //2
    return binary_str if binary_str else "0"

print(binary(1))
print(binary(5))
print(binary(10))

1 
101 
1010 


## Program 97
**Create a function that takes three arguments a, b, c and returns the sum of the
numbers that are evenly divided by c from the range a, b inclusive.**

Examples

evenly_divisible(1, 10, 20) ➞ 0

No number between 1 and 10 can be evenly divided by 20.

evenly_divisible(1, 10, 2) ➞ 30

2 + 4 + 6 + 8 + 10 = 30

evenly_divisible(1, 10, 3) ➞ 18

3 + 6 + 9 = 18


In [12]:
def evenly_divisible (a,b,c):
    total = 0
    for num in range(a,b+1):
        if num%c==0:
            total +=num
    return total

print(evenly_divisible(1, 10, 20))
print(evenly_divisible(1, 10, 2))
print(evenly_divisible(1, 10, 3))



0
30
18


## Program 98
**Create a function that returns True if a given inequality expression is correct and
False otherwise**

#### Examples
    
correct_signs("3 < 7 < 11") ➞ True
    
correct_signs("13 > 44 > 33 <1") ➞ False
    
correct_signs("1 < 2 < 6 < 9 > 3") ➞ True


In [13]:
def correct_signs(expression):
    try:
        return eval(expression)
    except:
        return False


print(correct_signs("3 < 7 < 11"))
print(correct_signs("13 > 44 > 33 < 1"))
print(correct_signs("1 < 2 < 6 < 9 > 3"))


True
False
True


## Program 99

**Create a function that replaces all the vowels in a string with a specified character.**

Examples

replace_vowels("the aardvark", "#") ➞ "th# ##rdv#rk"

replace_vowels("minnie mouse", "?") ➞ "m?nn?? m??s?"

replace_vowels("shakespeare", "*") ➞ "shkspr*"

In [15]:
def replace_vowels(string,char):
    vowels='AEIOUaeiou'
    for vowel in vowels:
        string=string.replace(vowel,char)
    return string

print(replace_vowels("the aardvark", "#"))
print(replace_vowels("minnie mouse", "?"))
print(replace_vowels("shakespeare", "*"))

th# ##rdv#rk
m?nn?? m??s?
sh*k*sp**r*


## Program 100
**Write a function that calculates the factorial of a number recursively.**

Examples

factorial(5) ➞ 120

factorial(3) ➞ 6

factorial(1) ➞ 1

factorial(0) ➞ 1

In [17]:
def factorial(n):
    if n==0:
        return 1
    else:
        return n* factorial(n-1)


print(factorial(12))
print(factorial(6))
print(factorial(5)) 
print(factorial(0))


479001600
720
120
1


## Program 101
**Hamming distance is the number of characters that differ between two strings.
To illustrate:**
    
String1: "abcbba"
    
String2: "abcbda"

Hamming Distance: 1 - "b" vs. "d" is the only difference.

Create a function that computes the hamming distance between two strings.

Examples

hamming_distance("abcde", "bcdef") ➞ 5

hamming_distance("abcde", "abcde") ➞ 0

hamming_distance("strong", "strung") ➞ 1


In [19]:
def hamming_distance(str1, str2):
    if len(str1) !=len(str2):
        raise ValueError("Input strings must have the same length")

    distance =0

    for i in range(len(str1)):
        if str1[i] != str2[i]:
            distance +=1

    return distance

print(hamming_distance("abcde","bcdef"))
print(hamming_distance("abcde","abcde"))
print(hamming_distance("strong","strung"))

5
0
1


## Program 102
**Create a function that takes a list of non-negative integers and strings and return a
new list without the strings.**

Examples

filter_list([1, 2, "a", "b"]) ➞ [1, 2]

filter_list([1, "a", "b", 0, 15]) ➞ [1, 0, 15]

filter_list([1, 2, "aasf", "1", "123", 123]) ➞ [1, 2, 123]


In [23]:
def filter_list(lst):
    result = []

    for element in lst:
        if isinstance(element,int) and element >=0:
            result.append(element)
    return result


In [24]:
filter_list([1, 2, "a", "b"]) 

[1, 2]

In [25]:
filter_list([1, "a", "b", 0, 15])

[1, 0, 15]

In [26]:
filter_list([1, 2, "aasf", "1", "123", 123])


[1, 2, 123]

## Program 103
**The "Reverser" takes a string as input and returns that string in reverse order, with
the opposite case.**

Examples

reverse("Hello World") ➞ "DLROw OLLEh"

reverse("ReVeRsE") ➞ "eSrEvEr"

reverse("Radar") ➞ "RADAr"

In [27]:
def reverse(input_str):
    reversed_str = input_str[::-1].swapcase()

    return reversed_str

In [28]:
reverse("Hello World")


'DLROw OLLEh'

In [29]:
reverse("ReVeRsE")

'eSrEvEr'

In [30]:
reverse("Radar") 

'RADAr'

## Program 104
**You can assign variables from lists like this:**

lst = [1, 2, 3, 4, 5, 6]

first = lst[0]

middle = lst[1:-1]

last = lst[-1]

print(first) ➞ outputs 1

print(middle) ➞ outputs [2, 3, 4, 5]

print(last) ➞ outputs 6

**With Python 3, you can assign variables from lists in a much more succinct way.
Create variables first, middle and last from the given list using destructuring
assignment (check the Resources tab for some examples), where:**

first ➞ 1

middle ➞ [2, 3, 4, 5]

last ➞ 6

In [32]:
writeyourcodehere = [1,2,3,4,5,6]
first, *middle, last = writeyourcodehere

In [36]:
print(first)
print(middle)
print(last)

1
[2, 3, 4, 5]
6


## Program 105
**Write a function that calculates the factorial of a number recursively.**

Examples

factorial(5) ➞ 120

factorial(3) ➞ 6

factorial(1) ➞ 1

factorial(0) ➞ 1


In [39]:
def factorial(n):
    if n==0:
        return 1
    else:
        return n*factorial(n-1)

In [42]:
print(factorial(5))
print(factorial(1))
print(factorial(0))
print(factorial(3))

120
1
1
6


## Program 106
**Write a function that moves all elements of one type to the end of the list.**

Examples

move_to_end([1, 3, 2, 4, 4, 1], 1) ➞ [3, 2, 4, 4, 1, 1]

Move all the 1s to the end of the array.

move_to_end([7, 8, 9, 1, 2, 3, 4], 9) ➞ [7, 8, 1, 2, 3, 4, 9]

move_to_end(["a", "a", "a", "b"], "a") ➞ ["b", "a", "a", "a"]


In [43]:
def move_to_end(lst,element):
    count=lst.count(element)
    lst=[x for x in lst if x!= element]

    lst.extend([element]*count)
    return lst

In [44]:
print(move_to_end([1, 3, 2, 4, 4, 1], 1))
print(move_to_end([1, 3, 2, 4, 4, 1], 1))
print(move_to_end(["a", "a", "a", "b"], "a") )

[3, 2, 4, 4, 1, 1]
[3, 2, 4, 4, 1, 1]
['b', 'a', 'a', 'a']


## Program 107
**Question1**

**Create a function that takes a string and returns a string in which each character is
repeated once.**
    
**Examples**
    
double_char("String") ➞ "SSttrriinngg"
    
double_char("Hello World!") ➞ "HHeelllloo WWoorrlldd!!"
    
double char("1234! ") ➞ "11223344!! "


In [45]:
def double_char(input_str):
    doubled_str = ""

    for char in input_str:
        doubled_str += char*2
    return doubled_str

In [46]:
double_char("string")

'ssttrriinngg'

In [47]:
double_char('deepak')

'ddeeeeppaakk'

In [48]:
double_char('Nisha')

'NNiisshhaa'

## Program 108
**Create a function that reverses a boolean value and returns the string "boolean
expected" if another variable type is given.**

Examples

reverse(True) ➞ False

reverse(False) ➞ True

reverse(0) ➞ "boolean expected"

reverse(None) ➞ "boolean expected"


In [50]:
def reverse(value):
    if isinstance(value, bool):
        return not value
    else:
        return "boolean excepted"
reverse(True)

False

In [51]:
reverse(False)

True

In [53]:
reverse(0)

'boolean excepted'

In [54]:
reverse(1)

'boolean excepted'

## Program 109
**Create a function that returns the thickness (in meters) of a piece of paper after
folding it n number of times. The paper starts off with a thickness of 0.5mm.**
    
Examples
    
num_layers(1) ➞ "0.001m"
    
- Paper folded once is 1mm (equal to 0.001m)
    
num_layers(4) ➞ "0.008m"
    
- Paper folded 4 times is 8mm (equal to 0.008m)
    
num_layers(21) ➞ "1048.576m"

- Paper folded 21 times is 1048576mm (equal to 1048.576m)


In [56]:
def num_layers(n):
    initial_thickness_mm = 0.5
    final_thickness_mm = initial_thickness_mm * (2**n)
    final_thickness_m =  final_thickness_mm / 1000
    return f"{final_thickness_m:.3f}m"

In [57]:
num_layers(1)

'0.001m'

In [58]:
num_layers(4)

'0.008m'

In [59]:
num_layers(21)

'1048.576m'

In [60]:
num_layers(34)

'8589934.592m'

## Program 110

**Create a function that takes a single string as argument and returns an ordered list
containing the indices of all capital letters in the string.**

Examples
    
index_of_caps("eDaBiT") ➞ [1, 3, 5]
    
index_of_caps("eQuINoX") ➞ [1, 3, 4, 6]
    
index_of_caps("determine") ➞ []
    
index_of_caps("STRIKE") ➞ [0, 1, 2, 3, 4, 5]
    
index_of_caps("sUn") ➞ [1]



In [64]:
def index_of_caps(word):
    return [i for i,char in enumerate(word) if char.isupper()]

In [65]:
index_of_caps("eDaBiT")


[1, 3, 5]

In [66]:
index_of_caps("eQuINoX") 

[1, 3, 4, 6]

In [67]:
index_of_caps("determine")


[]

## Program 111
**Using list comprehensions, create a function that finds all even numbers from 1 to
the given number.**

Examples

find_even_nums(8) ➞ [2, 4, 6, 8]

find_even_nums(4) ➞ [2, 4]

find_even_nums(2) ➞ [2]


In [68]:
def find_even_nums(num):
    return [x for x in range(1,num+1) if x % 2 ==0]

In [69]:
find_even_nums(8)

[2, 4, 6, 8]

In [70]:
find_even_nums(4)

[2, 4]

In [73]:
find_even_nums(90 )

[2,
 4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90]

## Program 112
**Create a function that takes a list of strings and integers, and filters out the list so
that it returns a list of integers only.**

Examples

filter_list([1, 2, 3, "a", "b", 4]) ➞ [1, 2, 3, 4]

filter_list(["A", 0, "Edabit", 1729, "Python", 1729]) ➞ [0, 1729]

filter_list(["Nothing", "here"]) ➞ []


In [74]:
def filter_list (lst):
    return [ x for x in lst if isinstance(x,int)]

In [75]:
filter_list([1, 2, 3, "a", "b", 4]) 

[1, 2, 3, 4]

In [76]:
filter_list(["A", 0, "Edabit", 1729, "Python", 1729]) 

[0, 1729, 1729]

In [77]:
filter_list(["A", 0, "Edabit", 1729, "Python", 1729])


[0, 1729, 1729]

In [78]:
filter_list(["Nothing", "here"])


[]

## Program 113
**Given a list of numbers, create a function which returns the list but with each
element's index in the list added to itself. This means you add 0 to the number at
index 0, add 1 to the number at index 1, etc...**
    
Examples
    
add_indexes([0, 0, 0, 0, 0]) ➞ [0, 1, 2, 3, 4]

add_indexes([1, 2, 3, 4, 5]) ➞ [1, 3, 5, 7, 9]

add_indexes([5, 4, 3, 2, 1]) ➞ [5, 5, 5, 5, 5]


In [79]:
def add_indexes(lst):
    return [i + val for i, val in enumerate(lst)]

In [80]:
add_indexes([0, 0, 0, 0, 0]) 


[0, 1, 2, 3, 4]

In [81]:
add_indexes([1, 2, 3, 4, 5])

[1, 3, 5, 7, 9]

In [82]:
add_indexes([5, 4, 3, 2, 1])

[5, 5, 5, 5, 5]

## Program 114
**Create a function that takes the height and radius of a cone as arguments and returns
the volume of the cone rounded to the nearest hundredth. See the resources tab for
the formula.**
                                                                              
Examples
                                                                              
cone_volume(3, 2) ➞ 12.57
                                                                              
cone_volume(15, 6) ➞ 565.49
                                                                              
cone_volume(18, 0) ➞ 0

In [83]:
import math

def cone_volume(height,radius):
    if radius ==0:
        return 0

    volume = (1/3) * math.pi * (radius**2) * height
    return round(volume , 2)

In [84]:
cone_volume(3, 2)

12.57

In [85]:
cone_volume(15 , 6)

565.49

In [87]:
cone_volume(18,0)

0

## Program 115
**This Triangular Number Sequence is generated from a pattern of dots that form a
triangle. The first 5 numbers of the sequence, or dots, are:**
    
1, 3, 6, 10, 15
    
**This means that the first triangle has just one dot, the second one has three dots, the
third one has 6 dots and so on.**
    
**Write a function that gives the number of dots with its corresponding triangle number
of the sequence.**
    
Examples
    
triangle(1) ➞ 1
    
triangle(6) ➞ 21
    
triangle(215) ➞ 23220

In [90]:
def triangle(n):
    if n < 1:
        return 0
    return n*(n+1)//2

In [91]:
triangle(1)

1

In [92]:
triangle(10)

55

In [93]:
trinagle(100)

5050

In [94]:
trinagle(215)

23220

## Program 116
**Create a function that takes a list of numbers between 1 and 10 (excluding one
number) and returns the missing number.**

Examples

missing_num([1, 2, 3, 4, 6, 7, 8, 9, 10]) ➞ 5

missing_num([7, 2, 3, 6, 5, 9, 1, 4, 8]) ➞ 10

missing_num([10, 5, 1, 2, 4, 6, 8, 3, 9]) ➞ 7


In [95]:
def missing_num(lst):
    total_sum = sum(range(1,11))
    given_sum = sum(lst)
    missing = total_sum - given_sum
    return missing

In [96]:
missing_num([1, 2, 3, 4, 6, 7, 8, 9, 10])

5

In [97]:
missing_num([7, 2, 3, 6, 5, 9, 1, 4, 8])


10

In [98]:
missing_num([10, 5, 1, 2, 4, 6, 8, 3, 9])


7

## Program 117
**Write a function that takes a list and a number as arguments. Add the number to the
end of the list, then remove the first element of the list. The function should then
return the updated list.**

Examples

next_in_line([5, 6, 7, 8, 9], 1) ➞ [6, 7, 8, 9, 1]

next_in_line([7, 6, 3, 23, 17], 10) ➞ [6, 3, 23, 17, 10]

next_in_line([1, 10, 20, 42 ], 6) ➞ [10, 20, 42, 6]

next_in_line([], 6) ➞ "No list has been selected"

In [99]:
def next_in_line(lst,num):
    if lst:
        lst.pop(0)
        lst.append(num)
        return lst
    else:
        return "No list has been selected"

In [100]:
next_in_line([5, 6, 7, 8, 9], 1) 



[6, 7, 8, 9, 1]

In [101]:
next_in_line([7, 6, 3, 23, 17], 10)


[6, 3, 23, 17, 10]

In [102]:
next_in_line([1, 10, 20, 42 ], 6)


[10, 20, 42, 6]

In [103]:
next_in_line([], 6)


'No list has been selected'

## Program 118
**Create the function that takes a list of dictionaries and returns the sum of people's
budgets.**

Examples

get_budgets([
    
{ 'name': 'John', 'age': 21, 'budget': 23000 },
    
{ 'name': 'Steve', 'age': 32, 'budget': 40000 },
    
{ 'name': 'Martin', 'age': 16, 'budget': 2700 }
    
]) ➞ 65700

get_budgets([
    
{ 'name': 'John', 'age': 21, 'budget': 29000 },
    
{ 'name': 'Steve', 'age': 32, 'budget': 32000 },
    
{ 'name': 'Martin', 'age': 16, 'budget': 1600 }
    1m
]) ➞ 62600


In [107]:
def get_budget(lst):
    total_budget = sum(person['budget'] for person in lst)
    return total_budget

budgets1 = [
 {'name': 'John', 'age': 21, 'budget': 23000},
 {'name': 'Steve', 'age': 32, 'budget': 40000},
 {'name': 'Martin', 'age': 16, 'budget': 2700}
]
budgets2 = [
 {'name': 'John', 'age': 21, 'budget': 29000},
 {'name': 'Steve', 'age': 32, 'budget': 32000},
 {'name': 'Martin', 'age': 16, 'budget': 1600}
]

In [108]:
get_budget(budgets1)

65700

In [110]:
get_budget(budgets2)

62600

## Program 119
**Create a function that takes a string and returns a string with its letters in
alphabetical order.**

Examples

alphabet_soup("hello") ➞ "ehllo"

alphabet_soup("edabit") ➞ "abdeit"

alphabet_soup("hacker") ➞ "acehkr"

alphabet_soup("geek") ➞ "eegk"

alphabet_soup("javascript") ➞ "aacijprstv"

In [111]:
def alphabet_soup(txt):
    return''.join(sorted(txt))

In [112]:
alphabet_soup("hello") 


'ehllo'

In [113]:
alphabet_soup("Deepak") 


'Daeekp'

In [114]:
alphabet_soup("Nisha") 


'Nahis'

In [115]:
alphabet_soup("javascript")

'aacijprstv'

## Program 120
**Suppose that you invest $10,000 for 10 years at an interest rate of 6% compounded
monthly. What will be the value of your investment at the end of the 10 year period?
Create a function that accepts the principal p, the term in years t, the interest rate r,
and the number of compounding periods per year n. The function returns the value at
the end of term rounded to the nearest cent.**

For the example:

compound_interest(10000, 10, 0.06, 12) ➞ 18193.97

Note that the interest rate is given as a decimal and n=12 because with monthly

compounding there are 12 periods per year. Compounding can also be done

annually, quarterly, weekly, or daily.

Examples

compound_interest(100, 1, 0.05, 1) ➞ 105.0

compound_interest(3500, 15, 0.1, 4) ➞ 15399.26

compound_interest(100000, 20, 0.15, 365) ➞ 2007316.26

In [116]:
def compound_interest(p,t,r,n):
    a = p *(1+(r/n)) ** (n*t)
    return round(a,2)

In [117]:
compound_interest(10000, 10, 0.06, 12)


18193.97

In [118]:
compound_interest(100, 1, 0.05, 1)


105.0

In [119]:
compound_interest(3500, 15, 0.1, 4)

15399.26

In [120]:
compound_interest(100000, 20, 0.15, 365)


2007316.26

## Program 121
**Write a function that takes a list of elements and returns only the integers.**

Examples

return_only_integer([9, 2, "space", "car", "lion", 16]) ➞ [9, 2, 16]

return_only_integer(["hello", 81, "basketball", 123, "fox"]) ➞ [81, 123]

return_only_integer([10, "121", 56, 20, "car", 3, "lion"]) ➞ [10, 56, 20,3]

return_only_integer(["String", True, 3.3, 1]) ➞ [1]

In [133]:
def return_only_integer(lst):
    return [x for x in lst if isinstance(x,int)  and not isinstance(x,bool)]

In [134]:
return_only_integer([9, 2, "space", "car", "lion", 16])

[9, 2, 16]

In [135]:
return_only_integer(["hello", 81, "basketball", 123, "fox"]) 


[81, 123]

In [136]:
return_only_integer([10, "121", 56, 20, "car", 3, "lion"]) 


[10, 56, 20, 3]

In [137]:
return_only_integer(["String", True, 3.3, 1])


[1]

## Program 122

**Create a function that takes three parameters where:**
    
- x is the start of the range (inclusive).

- y is the end of the range (inclusive).

- n is the divisor to be checked against.

Return an ordered list with numbers in the range that are divisible by the third

parameter n.

Return an empty list if there are no numbers that are divisible by n.
    
Examples
    
list_operation(1, 10, 3) ➞ [3, 6, 9]
    
list_operation(7, 9, 2) ➞ [8]
    
list_operation(15, 20, 7) ➞ []

In [139]:
def list_operation(x,y,n):
    return [num for num in range(x,y+1) if num % n ==0]

In [141]:
list_operation(1,10,2)

[2, 4, 6, 8, 10]

In [142]:
list_operation(1, 10, 3)


[3, 6, 9]

In [143]:
list_operation(15, 20, 7)


[]

## Program 123
**Create a function that takes in two lists and returns True if the second list follows the
first list by one element, and False otherwise. In other words, determine if the second
list is the first list shifted to the right by 1.**

Examples

simon_says([1, 2], [5, 1]) ➞ True

simon_says([1, 2], [5, 5]) ➞ False

simon_says([1, 2, 3, 4, 5], [0, 1, 2, 3, 4]) ➞ True

simon_says([1, 2, 3, 4, 5], [5, 5, 1, 2, 3]) ➞ False

Notes:

- Both input lists will be of the same length, and will have a minimum length of 2.

- The values of the 0-indexed element in the second list and the n-1th indexed

element in the first list do not matter.


In [148]:
def simon_says(list1,list2):
    return list1[:-1] == list2[1:]

In [149]:
simon_says([1,2],[5,1])

True

In [150]:
simon_says([1, 2], [5, 5])


False

In [151]:
simon_says([1, 2, 3, 4, 5], [0, 1, 2, 3, 4])


True

In [152]:
simon_says([1, 2, 3, 4, 5], [5, 5, 1, 2, 3])

False

## Program 124
**A group of friends have decided to start a secret society. The name will be the first
letter of each of their names, sorted in alphabetical order. Create a function that takes
in a list of names and returns the name of the secret society.**
    
Examples
    
society_name(["Adam", "Sarah", "Malcolm"]) ➞ "AMS"
    
society_name(["Harry", "Newt", "Luna", "Cho"]) ➞ "CHLN"
    
society_name(["Phoebe", "Chandler", "Rachel", "Ross", "Monica", "Joey"])


In [1]:
def society_name(names):
    secret_name = ''.join(sorted([name[0] for name in names]))
    return secret_name

In [2]:
society_name(["Adam", "Sarah", "Malcolm"])

'AMS'

In [3]:
society_name(["Harry", "Newt", "Luna", "Cho"])

'CHLN'

In [4]:
society_name(["Phoebe", "Chandler", "Rachel", "Ross", "Monica", "Joey"])

'CJMPRR'

## Program 125
**An isogram is a word that has no duplicate letters. Create a function that takes a
string and returns either True or False depending on whether or not it's an "isogram"**

Examples

is_isogram("Algorism") ➞ True

is_isogram("PasSword") ➞ False

- Not case sensitive.
    
is_isogram("Consecutive") ➞ False

Notes

Ignore letter case (should not be case sensitive).

All test cases contain valid one word strings.


In [8]:
def is_isogram(word):
    word = word.lower()

    unique_letters = set ()

    for letter in word:
        if letter in unique_letters:
            return False

        unique_letters.add(letter)
    return True

In [9]:
is_isogram("Algorism")


True

In [10]:
is_isogram("PasSword") 

False

In [11]:
is_isogram("Consecutive")


False

## Program 126
**Create a function that takes a string and returns True or False, depending on whether
the characters are in order or not.**
    
Examples
    
is_in_order("abc") ➞ True

is_in_order("edabit") ➞ False

is_in_order("123") ➞ True

is_in_order("xyzz") ➞ True

Notes

You don't have to handle empty strings.

In [12]:
def is_in_order(s):

        return s == ''.join(sorted(s))

In [13]:
is_in_order("abc") 


True

In [14]:
is_in_order("edabit")

False

In [15]:
is_in_order("xyzz")

True

## Program 127
**Create a function that takes a number as an argument and returns True or False
depending on whether the number is symmetrical or not. A number is symmetrical
when it is the same as its reverse.**

Examples

is_symmetrical(7227) ➞ True

is_symmetrical(12567) ➞ False

is_symmetrical(44444444) ➞ True

is_symmetrical(9939) ➞ False

is_symmetrical(1112111) ➞ True

In [16]:
def is_symmetrical(num):
    num_str = str(num)

    return num_str == num_str[::-1]

In [17]:
is_symmetrical(12567)

False

In [18]:
is_symmetrical(44444444)


True

In [19]:
is_symmetrical(44444444) 


True

In [20]:
is_symmetrical(1112111)

True

## Program 128
**Given a string of numbers separated by a comma and space, return the product of
the numbers.**
    
Examples
    
multiply_nums("2, 3") ➞ 6
    
multiply_nums("1, 2, 3, 4") ➞ 24
    
multiply_nums("54, 75, 453, 0") ➞ 0
    
multiply_nums("10, -2") ➞ -20


In [42]:
def multiply_nums(nums_str):
 # Split the input string by comma and space, then convert to intege
    nums = [int(num) for num in nums_str.split(", ")]
 
 # Initialize the result with 1
    result = 1
 
 # Multiply all the numbers together
    for num in nums:
        result *= num
 
    return result
    

In [43]:
multiply_nums("2, 3") 

6

In [44]:
multiply_nums("1, 2, 3, 4")

24

In [45]:
multiply_nums("54, 75, 453, 0") 

0

In [46]:
multiply_nums("10, -2")


-20

## Program 129

**Create a function that squares every digit of a number.**
    
Examples
    
square_digits(9119) ➞ 811181
    
square_digits(2483) ➞ 416649
    
square_digits(3212) ➞ 9414
    
Notes
    
The function receives an integer and must return an integer.

In [47]:
def square_digits(n):
    num_str = str (n)
    result_str = ""

    for digit in num_str:
        squared_digit = int(digit) ** 2
        result_str += str(squared_digit)
    return int(result_str)

In [48]:
square_digits(9119) 

811181

In [49]:
square_digits(2483)

416649

In [50]:
square_digits(3212) 

9414

## Program 130

**Create a function that sorts a list and removes all duplicate items from it.**

Examples

setify([1, 3, 3, 5, 5]) ➞ [1, 3, 5]

setify([4, 4, 4, 4]) ➞ [4]

setify([5, 7, 8, 9, 10, 15]) ➞ [5, 7, 8, 9, 10, 15]

setify([3, 3, 3, 2, 1]) ➞ [1, 2, 3]

In [51]:
def setify(lst):
    unique_set = set(sorted(lst))

    return list(unique_set)

In [52]:
setify([1, 3, 3, 5, 5])

[1, 3, 5]

In [53]:
setify([4, 4, 4, 4]) 

[4]

In [54]:
setify([3, 3, 3, 2, 1])

[1, 2, 3]

In [55]:
setify([5, 7, 8, 9, 10, 15])

[5, 7, 8, 9, 10, 15]

## Program 131

**Create a function that returns the mean of all digits.**
    
Examples
    
mean(42) ➞ 3
    
mean(12345) ➞ 3
    
mean(666) ➞ 6
    
Notes
    
The mean of all digits is the sum of digits / how many digits there are (e.g. mean
                                                                         
of digits in 512 is (5+1+2)/3(number of digits) = 8/3=2).

The mean will always be an integer.


In [56]:
def mean(n):
    n_str = str(n)
    digit_sum = sum (int(digit) for digit in n_str)

    digit_count = len(n_str)
    digit_mean = digit_sum/digit_count

    return int(digit_mean)

In [57]:
mean(42)

3

In [58]:
mean(123456778)

4

## Program 132
**Create a function that takes an integer and returns a list from 1 to the given number,
where:**
    
1. If the number can be divided evenly by 4, amplify it by 10 (i.e. return 10 times the

number).

2. If the number cannot be divided evenly by 4, simply return the number.

Examples

amplify(4) ➞ [1, 2, 3, 40]
    
amplify(3) ➞ [1, 2, 3]
    
amplify(25) ➞ [1, 2, 3, 40, 5, 6, 7, 80, 9, 10, 11, 120, 13, 14, 15, 160, 17, 18, 19, 200, 21,
22, 23, 240, 25]
    
Notes

The given integer will always be equal to or greater than 1.
    
Include the number (see example above).

To perform this problem with its intended purpose, try doing it with list

comprehensions. If that's too difficult, just solve the challenge any way you can.


In [59]:
def amplify (num):
    return [n * 10 if n%4 == 0 else n for n in range(1,num+1)]

In [60]:
amplify(4)

[1, 2, 3, 40]

In [61]:
amplify(3)

[1, 2, 3]

In [62]:
amplify(20)

[1, 2, 3, 40, 5, 6, 7, 80, 9, 10, 11, 120, 13, 14, 15, 160, 17, 18, 19, 200]

In [63]:
amplify(25)

[1,
 2,
 3,
 40,
 5,
 6,
 7,
 80,
 9,
 10,
 11,
 120,
 13,
 14,
 15,
 160,
 17,
 18,
 19,
 200,
 21,
 22,
 23,
 240,
 25]

## Program 133
**Create a function that takes a list of numbers and return the number that's unique.**

Examples

unique([3, 3, 3, 7, 3, 3]) ➞ 7

unique([0, 0, 0.77, 0, 0]) ➞ 0.77

unique([0, 1, 1, 1, 1, 1, 1, 1]) ➞ 0

Notes

Test cases will always have exactly one unique number while all others are the same.

In [64]:
def unique(numbers):
    count_dict = {}
    for num in numbers:
        if num in count_dict:
            count_dict[num] +=1
        else:
            count_dict[num] =1

    for num, count in count_dict.items():
        if count ==1:
            return num

In [65]:
unique([3, 3, 3, 7, 3, 3])

7

In [66]:
unique([0, 0, 0.77, 0, 0]) 


0.77

In [67]:
unique([0, 1, 1, 1, 1, 1, 1, 1])


0

## Program 134
**Your task is to create a Circle constructor that creates a circle with a radius provided
by an argument. The circles constructed must have two getters getArea() (PIr^2) and
getPerimeter() (2PI*r) which give both respective areas and perimeter
(circumference).**

**For help with this class, I have provided you with a Rectangle constructor which you
can use as a base example.**

Examples

circy = Circle(11)

circy.getArea()

Should return 380.132711084365

circy = Circle(4.44)

circy.getPerimeter()

Should return 27.897342763877365

Notes

Round results up to the nearest integer.

In [72]:
import math
class circle:
    def __init__(self,radius):
        self.radius = radius
    def getArea(self):
        return round(math.pi*self.radius**2)

    def getPerimeter(self):
        return round(2*math.pi * self.radius)

In [74]:
circy = circle(11)
print(circy.getArea()) 
print(circy.getPerimeter())


380
69


In [76]:
circy = circle(4.44)
print(circy.getArea()) 
print(circy.getPerimeter())


62
28


## Program 135
**Create a function that takes a list of strings and return a list, sorted from shortest to
longest.**

Examples

sort_by_length(["Google", "Apple", "Microsoft"]) ➞ ["Apple", "Google", "Microsoft"]

sort_by_length(["Leonardo", "Michelangelo", "Raphael", "Donatello"]) ➞ ["Raphael",

"Leonardo", "Donatello", "Michelangelo"]

sort_by_length(["Turing", "Einstein", "Jung"]) ➞ ["Jung", "Turing", "Einstein"]

Notes

All test cases contain lists with strings of different lengths, so you won't have to deal

In [79]:
def sort_by_length(lst):
    return sorted(lst,key =len)

In [80]:
sort_by_length(["Google", "Apple", "Microsoft"]) 

['Apple', 'Google', 'Microsoft']

In [81]:
sort_by_length(["Leonardo", "Michelangelo", "Raphael", "Donatello"])

['Raphael', 'Leonardo', 'Donatello', 'Michelangelo']

In [82]:
sort_by_length(["Turing", "Einstein", "Jung"])

['Jung', 'Turing', 'Einstein']

## Program 136
**Create a function that validates whether three given integers form a Pythagorean
triplet. The sum of the squares of the two smallest integers must equal the square of
the largest number to be validated.**

Examples

is_triplet(3, 4, 5) ➞ True

3² + 4² = 25

5² = 25

is_triplet(13, 5, 12) ➞ True

5² + 12² = 169

13² = 169

is_triplet(1, 2, 3) ➞ False

1² + 2² = 5

3² = 9

Notes

Numbers may not be given in a sorted order.


In [85]:
def triplet(a,b,c):
    sorted_numbers = sorted([a,b,c])
    return sorted_numbers[0] ** 2 + sorted_numbers[1] ** 2 == sorted_numbers

In [86]:
triplet(3, 4, 5)

False

In [87]:
triplet(13, 5, 12) 

False

In [88]:
triplet(1, 2, 3)

False

## Program 137
**Create a function that takes three integer arguments (a, b, c) and returns the amount
of integers which are of equal value.**

Examples

equal(3, 4, 3) ➞ 2

equal(1, 1, 1) ➞ 3

equal(3, 4, 1) ➞ 0

Notes

Your function must return 0, 2 or 3

In [89]:
def equal(a,b,c):
    if a==b==c:
        return 3
    elif a==b or b==c or a==c:
        return 2
    else:
        return 0

In [90]:
equal(3, 4, 3)


2

In [91]:
equal(1, 1, 1) 

3

In [92]:
equal(3, 4, 1)

0

## Program 138
**Write a function that converts a dictionary into a list of keys-values tuples.**
    
Examples
    
dict_to_list({
    
"D": 1,

"B": 2
    
"C": 3
    
}) ➞ [("B", 2), ("C", 3), ("D", 1)]
    
dict_to_list({
    
"likes": 2,
    
"dislikes": 3,
    
"followers": 10
    
}) ➞ [("dislikes", 3), ("followers", 10), ("likes", 2)]
    
Notes
    
Return the elements in the list in alphabetical order.


In [93]:
def  dict_to_list(input_dict):
    sorted_dict = sorted(input_dict.items())

    result = [(key , value) for key, value in sorted_dict]
    return result

In [94]:
dict_to_list({
 "D": 1, 
 "B": 2, 
 "C": 3
})

[('B', 2), ('C', 3), ('D', 1)]

In [95]:
dict_to_list({
"likes": 2,
"dislikes": 3,
"followers": 10
})

[('dislikes', 3), ('followers', 10), ('likes', 2)]

## Program 139
**Write a function that creates a dictionary with each (key, value) pair being the (lower
case, upper case) versions of a letter, respectively.**

Examples

mapping(["p", "s"]) ➞ { "p": "P", "s": "S" }

mapping(["a", "b", "c"]) ➞ { "a": "A", "b": "B", "c": "C" }

mapping(["a", "v", "y", "z"]) ➞ { "a": "A", "v": "V", "y": "Y", "z": "Z" }

Notes

All of the letters in the input list will always be lowercase.


In [96]:
def mapping(letters):
    result = {}
    for letter in letters :
        result[letter] = letter.upper()
    return result

In [97]:
mapping(["p", "s"]) 


{'p': 'P', 's': 'S'}

In [98]:
mapping(["a", "b", "c"])

{'a': 'A', 'b': 'B', 'c': 'C'}

In [99]:
mapping(["a", "v", "y", "z"]) 

{'a': 'A', 'v': 'V', 'y': 'Y', 'z': 'Z'}

## Program 140
**Write a function, that replaces all vowels in a string with a specified vowel.**

Examples

vow_replace("apples and bananas", "u") ➞ "upplus und bununus"

vow_replace("cheese casserole", "o") ➞ "chooso cossorolo"

vow_replace("stuffed jalapeno poppers", "e") ➞ "steffed jelepene peppers"

Notes

All words will be lowercase. Y is not considered a vowel.

In [100]:
def vowel_replace(string,vowel):
    vowels = 'aeiou'
    result=''
    for char in string:
        if char in vowels:
            result +=vowel
        else:
            result+= char

    return result

In [102]:
vowel_replace("apples and bananas", "u")

'upplus und bununus'

In [103]:
vowel_replace("cheese casserole", "o")


'chooso cossorolo'

In [104]:
vowel_replace("stuffed jalapeno poppers", "e") 


'steffed jelepene peppers'

## Program 141

**Create a function that takes a string as input and capitalizes a letter if its ASCII code
is even and returns its lower case version if its ASCII code is odd.**
    
Examples
    
ascii_capitalize("to be or not to be!") ➞ "To Be oR NoT To Be!"
    
ascii_capitalize("THE LITTLE MERMAID") ➞ "THe LiTTLe meRmaiD"
    
ascii_capitalize("Oh what a beautiful morning.") ➞ "oH wHaT a BeauTiFuL
    
moRNiNg."


In [105]:
def ascii_capitalize(input_str):
    result = ""

    for char in input_str:
        if ord(char) %2 ==0:
            result +=char.upper()
        else:
            result += char.lower()
    return result

In [106]:
ascii_capitalize("to be or not to be!") 

'To Be oR NoT To Be!'

In [107]:
ascii_capitalize("THE LITTLE MERMAID") 


'THe LiTTLe meRmaiD'

In [108]:
ascii_capitalize("Oh what a beautiful morning.") 

'oH wHaT a BeauTiFuL moRNiNg.'

In [19]:
import numpy as np
arr=[34455,'deepek']
a=np.array(arr)

In [20]:
type(a)

numpy.ndarray

In [59]:
l1=[2,3,4,5]
l2 = [x**2 for x in l1]
l2

[4, 9, 16, 25]

14

In [16]:
d1={'a':100,'b':200,'c':300}
d2={'a':300,'b':200,'d':400}

In [28]:
for i in range(min(len(d1),len(d2))):
    d1[list(d1.keys())[i]]+=d2[list(d2.keys())[i]]
d1

{'a': 400, 'b': 400, 'c': 700}

In [29]:
d1 = {'a': 100, 'c': 300,'b':900}
d2 = {'a': 300, 'b': 200, 'd': 400}

res = {}


for key, value in d1.items():
    res[key] = value


for key, value in d2.items():
    if key in res:
        res[key] += value
    else:
        res[key] = value

print(res)


{'a': 400, 'c': 300, 'b': 1100, 'd': 400}
